### ISSUE TITLE
Separate container images for primary task and subtasks in multi-container task

### ISSUE THREAD

**Issue by rytmis** [Azure/Batch#150]
I'm working on a setup where I've got:
- A .NET console application that runs a bunch of HTTP requests in parallel and aggregates the results. It can natively distribute the requests across a number of hosts.
- A stateless server (based on static data) that serves the HTTP requests, and should fanout to multiple compute nodes to enable the console app to complete its work as quickly as possible.

Multi-container tasks would seem like a good fit otherwise, but as I understand the docs, the primary application task and the fanout tasks should share a container image. This is currently not an option for me, mainly because the server image is based on an old Alpine image and can't support any current .NET release.

Is Azure Batch a feasible option for this scenario?

**Comment by alfpark** [comment]
Multi-instance tasks are for coordinated execution tasks. Does your task require all of the sub-tasks of the multi-instance task to execute concurrently to make forward progress? If not, you don't need to use multi-instance tasks.

You probably want [JobManager](https://learn.microsoft.com/azure/batch/jobs-and-tasks#job-manager-task) tasks instead?

**Comment by rytmis** [comment]
The way things are currently structured means that I only get a result if all the subtasks are performed, and doing it on a single node would take approximately three eternities. Hence my desire for a solution where I can run parallel servers and distribute the work. However, the parallel servers alone won’t do anything—they need the query application to run the queries.

I understand this is not a perfect fit for Batch, but I was hoping I could squint a bit and make it work. 

I’m not overly attached to any specific solution, so maybe Job Manager is the way here. If a job manager task starts other tasks, will it know the addresses of the nodes running those tasks? That would be necessary, or otherwise I’ll have to reorganize the solution. I’m not against doing that per se, but I would first like to know if I might not have to.

**Comment by rytmis** [comment]
The way things are currently structured means that I only get a result if all the subtasks are performed, and doing it on a single node would take approximately three eternities. Hence my desire for a solution where I can run parallel servers and distribute the work. However, the parallel servers alone won’t do anything—they need the query application to run the queries.

I understand this is not a perfect fit for Batch, but I was hoping I could squint a bit and make it work. 

I’m not overly attached to any specific solution, so maybe Job Manager is the way here. If a job manager task starts other tasks, will it know the addresses of the nodes running those tasks? That would be necessary, or otherwise I’ll have to reorganize the solution. I’m not against doing that per se, but I would first like to know if I might not have to.

**Comment by alfpark** [comment]
Sorry for the delay here.

Job Manager could do this via reverse lookups on the task -> compute node running the task. This isn't very elegant and can lead to a lot of code to cover corner cases.

I may not be 100% clear in understanding your scenario, but my recommendation would be to convert this into a work pulling system (e.g., create a queue for work where tasks would be generic to pull work from the queue when scheduled) or do some heavy lifting and convert your pipeline to use Batch jobs and tasks natively (i.e., push work as formal Batch tasks and let the system take care of the rest).

**Comment by rytmis** [comment]
No worries, I figured out a solution that worked for me in the meantime.

I agree that a work-pulling or partitioning solution would be ideal, and I may do so in the future. For now, I needed something that I could hack up a bit quicker.

What I ended up doing was this: I wrapped the console app with a PowerShell script that gets node IDs and IP addresses as env variables (batch_node_x_id, batch_node_x_ipaddr). It filters out the pair that matches AZ_BATCH_NODE_ID and uses the remaining list. It loops through the list, and for each individual IP address waits until it can get a successful response. This wastes a bit of time, but in practice every worker node is up and running in approximately a minute, so it doesn't really matter.

In the other end of things, the code that creates the job and tasks first scales the pool to the desired size, then gets a list of all the nodes and creates the environment variables from that list. Yes, this is very hacky, but this is a first stab at figuring out how big of an operation we are dealing with, so it doesn't need to be production grade in its fault tolerance or elegance. :)
